src/clj_rabin/hash.clj

```diff
@@ -1,51 +1,103 @@
+  (:import (clojure.lang PersistentVector)
+           (java.io BufferedInputStream InputStream)))
 
 (def default-ctx
   "prime              : should be close to the alphabet size
    q (modulus)        : should be sufficiently large to avoid collisions, also prime
    window-size (bytes): can be anything (but as little as 16 bytes is 'enough')"
-  {:prime       257
-   :q           153191                                      ;Integer/MAX_VALUE
-   :window-size 16})
+  {:prime       (long 257)
+   :q           (long 153191)                               ;Integer/MAX_VALUE
+   :window-size (int 32)})
 
-(defn bigint-pow
-  [a b]
-  (reduce * (repeat b (bigint a))))
+(defn mod-pow
+  ^long
+  [^long a ^long b ^long q]
+  (reduce (fn [a b]
+            (mod (* a b) q)) 1 (repeat b a)))
+
+(def range-vec
+  (memoize (fn ^PersistentVector
+             [s & [e]]
+             (into [] (if e
+                        (range s e)
+                        (range s))))))
 
 (defn lsb-zero?
   "Are the bottom n bits 0?"
-  [hash n]
-  (let [hash (int hash)]
-    (-> hash
-        (bit-shift-right n)
-        (bit-shift-left n)
-        (bit-and hash)
-        (= hash))))
-
-(defn window-pow
-  "pow = p^window-sz % q"
-  [{:keys [window-size prime q]}]
-  (-> (bigint-pow prime window-size)
-      (mod q)))
+  [^long hash ^Integer n]
+  (-> hash
+      (unsigned-bit-shift-right ^long n)
+      (bit-shift-left ^long n)
+      (bit-and ^long hash)
+      (= hash)))
 
 (defn poly-hash
   "Compute the polynomial hash for a window
   P^w*a[i] + P^w-1[i+1] + ..."
-  [{:keys [window-size prime q]} ^bytes bs]
-  (let [hash (reduce (fn [acc i]
-                       (-> (bigint-pow prime (- (dec window-size) i))
-                           (* (nth bs i))
+  ^long
+  [{:keys [^long window-size ^long prime ^long q]} ^bytes bs]
+  (let [hash (reduce (fn [^long acc ^long i]
+                       (-> (mod-pow prime (- (dec window-size) i) q)
+                           (* ^byte (nth bs i))
                            (+ acc)))
-                     0
-                     (range window-size))]
+                     (long 0)
+                     (range-vec window-size))]
     (mod hash q)))
 
+(defn slide-roll-hash
+  "Increment the hash given the byte entering the window (in-byte)
+  and the byte leaving the window (out-byte)"
+  ^long
+  [{:keys [^long prime ^long pow ^long q]} ^long prev-hash out-byte in-byte]
+  (-> (* prev-hash prime)
+      (+ ^byte in-byte)
+      (- (* ^byte out-byte ^long pow))
+      (mod q)))
+
+(defn do-rabin
+  "Given a function (for side effects), a rabin context, and some bytes, emit a seq of
+  [[index rabin-hash] ...] each index beginning from the end of the first window
+
+  pow (computed for you) = p^window-sz % q
+
+  opts:
+
+  :window-size Sliding window size
+  :prime       Rabin Polynomial constant
+  :q           Modulus"
+  ([f ^bytes bs]
+   (do-rabin f default-ctx bs))
+  ([f ctx ^bytes bs]
```

Comment:
this feels like it'd fit better an a trailing opt map, specially since the examples you have at the bottom just ignore the input

```clj
(defn do-rabin
  [ctx ^bytes bs & {:keys [hash-fn] :as _opts}] ...)
```